### PR TITLE
Add (de)serialization for immutable lists

### PR DESCRIPTION
**Pull request by DrSmugleaf**
Also tests for it.
The reflection calls are a bit arse but I don't know of another way to cast it.

### CODE REVIEW

File: Robust.Shared/Serialization/YamlObjectSerializer.cs [hunk 1]
```diff
@@ -510,6 +511,22 @@ public object NodeToType(Type type, YamlNode node)
...
+            {
+                var listNode = (YamlSequenceNode) node;
+                var list = typeof(ImmutableList<>).MakeGenericType(immutableListType);
+                var immutableList = list.GetField("Empty")!.GetValue(null);
+                var add = list.GetMethod("Add")!;
+
+                foreach (var entryNode in listNode.Children)
+                {
+                    var value = NodeToType(immutableListType, entryNode);
+                    immutableList = add.Invoke(immutableList, new[] {value});
```
Comment (PJB3005) [inline Robust.Shared/Serialization/YamlObjectSerializer.cs]: Isn't this literally allocating a new list for every single element, resulting in O(n!) behavior?

You should probably copy the elements into an intermediate array or list and then create the immutablelist to finish it.

Comment (DrSmugleaf) [reply Robust.Shared/Serialization/YamlObjectSerializer.cs]: Good point it is, I'll fix it tomorrow then.

Comment (DrSmugleaf) [reply Robust.Shared/Serialization/YamlObjectSerializer.cs]: Fixed.

**Comment by DrSmugleaf** [comment]
Reviews addressed.

**Pull request merged**